allow selection of image project for gce

pbx0 opened the pull request:
The gcloud should distinguish the image project from the instance project. Mainly, its useful to be able to select vm images from a project like coreos-cloud but have the instances still spawned on the coreos-gce-testing project. Otherwise one has to guess which jenkin's created image is good. 

cc @marineam @mischief

marineam commented:
Alternatively, should we support longer form image paths that include the
project?

On Dec 19, 2016 5:00 PM, "Patrick Baxter" <notifications@github.com> wrote:

> The gcloud should distinguish the image project from the instance project.
> Mainly, its useful to be able to select vm images from a project like
> coreos-cloud but have the instances still spawned on the coreos-gce-testing
> project. Otherwise one has to guess which jenkin's instance is good.
>
> cc @marineam <https://github.com/marineam> @mischief
> <https://github.com/mischief>
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/coreos/mantle/pull/429
> Commit Summary
>
>    - platform/api/gcloud: distinguish image project from instance project
>    - cmd/kola: allow setting the gce image project
>
> File Changes
>
>    - *M* cmd/kola/options.go
>    <https://github.com/coreos/mantle/pull/429/files#diff-0> (3)
>    - *M* platform/api/gcloud/api.go
>    <https://github.com/coreos/mantle/pull/429/files#diff-1> (15)
>    - *M* platform/api/gcloud/compute.go
>    <https://github.com/coreos/mantle/pull/429/files#diff-2> (9)
>    - *M* platform/api/gcloud/image.go
>    <https://github.com/coreos/mantle/pull/429/files#diff-3> (10)
>
> Patch Links:
>
>    - https://github.com/coreos/mantle/pull/429.patch
>    - https://github.com/coreos/mantle/pull/429.diff
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/coreos/mantle/pull/429>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAJYKNLJo4KfywIj67kXW1hayR3EwqZHks5rJyikgaJpZM4LRWhe>
> .
>

pbx0 commented:
@marineam If we support longer-form image paths we could reduce `--gce-image` and `--gce-image-project` to one flag, but it would be a breaking change which I don't think is that big a deal for us.  Also we should change `--gce-project` to the long-form for parity between the flags.

If your OK with a minor breakage of jenkins jobs i'm happy to change to long-form.

marineam commented:
I meant support both, assume the project id when if the image name doesn't
contain the parent project path. Other CLI tools behave similarly.

That behavior wouldn't break anything would it?

On Dec 20, 2016 2:51 PM, "Patrick Baxter" <notifications@github.com> wrote:

> @marineam <https://github.com/marineam> If we support longer-form image
> paths we could reduce --gce-image and --gce-image-project to one flag,
> but it would be a breaking change which I don't think is that big a deal
> for us. Also we should change --gce-project to the long-form for parity
> between the flags.
>
> If your OK with a minor breakage of jenkins jobs i'm happy to change to
> long-form.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/coreos/mantle/pull/429#issuecomment-268382364>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAJYKFh0TF3jKqhfAZDtDi6lc5SJPuK1ks5rKFvmgaJpZM4LRWhe>
> .
>

pbx0 commented:
Oh, yes, that'd be fine. I will update to support long and short form names for the images.

pbx0 commented:
Updated so that the image project will be taken from the image name if it is longform and from the Project flag otherwise.

marineam commented on platform/api/gcloud/compute.go:
```diff
@@ -66,7 +73,7 @@ func (a *API) mkinstance(userdata, name string, keys []*agent.Key) *compute.Inst
 				InitializeParams: &compute.AttachedDiskInitializeParams{
 					DiskName:    name,
-					SourceImage: prefix + "/global/images/" + a.options.Image,
+					SourceImage: imagePrefix + "/global/images/" + a.options.Image,
```
If Image is the full path `https://www.googleapis.com/compute/v1/projects/projname/global/images/imgname` Won't this create URLs like `https://www.googleapis.com/compute/v1/projects/projname/global/images/imgname/global/images/https://www.googleapis.com/compute/v1/projects/projname/global/images/imgname`?

pbx0 replied:
Ah, right, this is wrong.

marineam commented on platform/api/gcloud/compute.go:
```diff
@@ -46,10 +46,17 @@ func (a *API) mkinstance(userdata, name string, keys []*agent.Key) *compute.Inst
+
+	// If long image name given for image, respect the image project
+	imagePrefix := instancePrefix
+	if strings.HasPrefix(a.options.Image, "https") {
```
I don't know about other tools off the top of my head but the old gcutil accepted image names relative to the API endpoint, e.g. `projects/name/global/images/foo`

So how about this to be as flexible as possible:
 - If prefix is the API endpoint `https://www.googleapis.com/compute/v1/` pass the value through verbatim.
 - If prefix is `projects/` append it to the URL endpoint and use it.
 - If image contains no `/` construct the appropriate URL using the default project.
 - Otherwise the name is invalid.

pbx0 replied:
SGTM

pbx0 commented:
@marineam PTAL. I moved the logic to a more appropriate place and implemented support for the full endpoint url, the shortened endpoint url starting with projects, and just the plain image name. Anything that doesn't start with those two given prefixes and contains "/" is rejected.
